snuba/datasets/discover.py

```diff
@@ -74,99 +69,31 @@ def detect_dataset(
-            .get_dataset_schemas()
-            .get_read_schema()
-            .get_data_source()
-        )
-
-    def get_prewhere_candidates(self) -> Sequence[str]:
-        if self.__table_source:
-            return self.__table_source.get_prewhere_candidates()
-        return []
-
-
-class DatasetSelector(QueryProcessor):
-    """
-    Switches the table source to either Events or Transactions, depending on
-    the detected dataset.
-    """
-
-    # TODO: Make this generic. It does not need to be discover specific.
-
-    def __init__(
-        self,
-        discover_source: RelationalSource,
-        events_columns: ColumnSet,
-        transactions_columns: ColumnSet,
-        post_processors: Mapping[str, Sequence[QueryProcessor]],
-    ) -> None:
-        self.__discover_source = discover_source
-        self.__events_columns = events_columns
-        self.__transactions_columns = transactions_columns
-        # These are the processors that need to run depending on the selected dataset.
-        # Adding them here instead of as top level processors running their own conditions
-        # to make the coupling explicit.
-        self._post_processors = post_processors
-
-    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
-        detected_dataset = detect_dataset(
-            query, self.__events_columns, self.__transactions_columns
-        )
-        source = query.get_data_source()
-        assert isinstance(source, DiscoverSource)
-        source.set_table_source(detected_dataset)
-        for processor in self._post_processors.get(detected_dataset, []):
-            processor.process_query(query, request_settings)
-
-
-class DiscoverSchema(Schema):
+class DiscoverQueryStorageSelector(QueryStorageSelector):
     def __init__(
-        self, table_source, common_columns, events_columns, transactions_columns
+        self, events_table: TableStorage, transactions_table: TableStorage
     ) -> None:
-        self.__common_columns = common_columns
-        self.__events_columns = events_columns
-        self.__transactions_columns = transactions_columns
-        self.__data_source = DiscoverSource(self.get_columns(), table_source)
-
-    def get_data_source(self) -> DiscoverSource:
-        return self.__data_source
-
-    def get_columns(self) -> ColumnSet:
-        return (
-            self.__common_columns + self.__events_columns + self.__transactions_columns
+        self.__events_table = events_table
+        self.__transactions_table = transactions_table
+
+    def select_storage(
+        self, query: Query, request_settings: RequestSettings
+    ) -> Storage:
+        table = detect_table(
+            query,
+            self.__events_table.get_schemas().get_read_schema().get_columns(),
+            self.__transactions_table.get_schemas().get_read_schema().get_columns(),
         )
+        return self.__events_table if table == EVENTS else self.__transactions_table
```

Comment:
This is so much nicer than it was before.

Should `detect_table` just be inlined into this method instead of being a separate function, or is it possible for `detect_table` to at least take the storages instead of the column sets? Translating the `EVENTS`/`TRANSACTIONS` constants to the storages seems like unnecessary indirection here.

Reply:
unfortunately it is not possible. It is called by column_expr still. We can inline it once column_expr is gone which depends on the AST

Reply:
Ugh, right.